### PR TITLE
Cleanup logical optimizer rules.

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

## Rationale for this change
While working on another PR. I noticed that, some of the Logical plan rules doesn't use approriate constructors when re-creating a LogicalPlan node. Because of this reason, during rules we have schema mismatches, that are unnoticed. This PR solves this problem.

To prevent schema mismatches, I use `try_new` API, instead of `try_new_with_schema` API. Because of this change we receive schema mismatch errors during LogicalPlan optimization(These were bugs, that were unnoticed, because of the use of wrong schema). This PR includes bug fixes that are the cause of schema mismatch.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/array.slt [hunk 1]
```diff
@@ -2465,7 +2465,7 @@ true
 query B
 select empty(make_array(NULL));
 ----
-true
+false
```
Comment (mustafasrepo) [inline datafusion/sqllogictest/test_files/array.slt:2468]: This is the test change we talked about.

File: datafusion/expr/src/built_in_function.rs [hunk 2]
```diff
@@ -315,6 +316,78 @@ fn function_to_name() -> &'static HashMap<BuiltinScalarFunction, &'static str> {
...
+        (DataType::Float32 | DataType::Float64, DataType::Float16) => lhs.clone(),
+        (DataType::Float64, DataType::Float32) => lhs.clone(),
+        // String
+        (DataType::Utf8, DataType::Utf8 | DataType::LargeUtf8) => rhs.clone(),
+        (DataType::LargeUtf8, DataType::Utf8) => lhs.clone(),
+        (DataType::LargeUtf8, DataType::LargeUtf8) => DataType::LargeUtf8,
+        (DataType::List(lhs_field), DataType::List(rhs_field)) => {
+            let field_type =
+                get_wider_type(lhs_field.data_type(), rhs_field.data_type())?;
+            assert_eq!(lhs_field.name(), rhs_field.name());
```
Comment (berkaysynnada) [inline datafusion/expr/src/built_in_function.rs:377]: Is panicking the desired behavior when they are not same?

File: datafusion/expr/src/type_coercion/binary.rs [hunk 3]
```diff
@@ -120,7 +120,7 @@ fn signature(lhs: &DataType, op: &Operator, rhs: &DataType) -> Result<Signature>
         }
         Operator::AtArrow
         | Operator::ArrowAt => {
-            array_coercion(lhs, rhs).map(Signature::uniform).ok_or_else(|| {
+            array_coercion(lhs, rhs).map(Signature::comparison).ok_or_else(|| {
```
Comment (berkaysynnada) [inline datafusion/expr/src/type_coercion/binary.rs:123]: I think you can add a comment on that point, ArrowAt is not exactly a comparison.

File: datafusion/optimizer/src/single_distinct_to_groupby.rs [hunk 4]
```diff
@@ -108,18 +108,30 @@ impl OptimizerRule for SingleDistinctToGroupBy {
...
-                            let alias_str = format!("group_alias_{i}");
-                            let alias_expr = group_expr.clone().alias(&alias_str);
-                            group_expr_alias
-                                .push((alias_str, schema.fields()[i].clone()));
+                            let (alias_expr, out_group_expr, original_name) =
+                                if let Expr::Column(_) = group_expr {
+                                    // For Column expressions we can use existing expression as is.
+                                    (group_expr.clone(), group_expr.clone(), None)
+                                } else {
+                                    // For complex expression write is as alias, to be able to refer
```
Comment (berkaysynnada) [inline datafusion/optimizer/src/single_distinct_to_groupby.rs:116]: Can you write the comment a little more clearly?

**Review by ozankabak — APPROVED** [review]
LGTM, let's send upstream.

**Comment by mustafasrepo** [comment]
Upstream [PR](https://github.com/apache/arrow-datafusion/pull/7919).

**Comment by ozankabak** [comment]
Closing, merged upstream.

**Pull request closed without merging**